runtime/src/chpl-gpu.c

```diff
@@ -389,6 +389,269 @@ void chpl_gpu_comm_get(c_sublocid_t dst_subloc, void *dst,
   }
 }
 
+void chpl_gpu_comm_get_strd(c_sublocid_t dst_subloc,
+                          void* dstaddr_arg, size_t* dststrides,
+                          c_nodeid_t srclocale, c_sublocid_t src_subloc,
+                          void* srcaddr_arg, size_t* srcstrides,
+                          size_t* count, int32_t stridelevels, size_t elemSize,
+                          int32_t commID, int ln, int32_t fn) {
+  // TODO: Re-use code from chpl-comm-strd-xfer.h instead of copying it here.
```

Comment:
Certainly refactoring this to reuse code would be preferrable.  Taking a look at a diff of this and the original code (`get_strd_common`), I think it's doable but not trivial. I understand your hesitence.

The differences I see include:

- what parameters get passed to the function (the original version takes `maxOutstandingXfers` and `yieldfn` while the `chpl_gpu` version takes `src_subloc`)
- the communication callback code is missing in the new version (as is a conditional call to `chpl_comm_wait_nb_some` at the end of the function)
- several calls to `strd_nb_helper` get replaced with calls to `chpl_gpu_comm_get`.

In your OP you mention such a refactoring would require macros or function pointers (presumably to factor out and dependency-inject those diffing pieces). Rather than doing that how would you feel about making some common uber-function that takes the parameters of both then adds some `if` statements where needed?

It'd be kind of ugly (and the extra `if`s might be a performance concern?). But perhaps it's the lesser evil compared to leaving it duplicated. (and unless you say your planning to work on it next I'm not really very optimistic that this will ever be refactored).

Anyway, if that suggestion sounds horrible, I think at a minimum I'd add a comment warning about the duplication on the `chpl-comm-strd-xfer.h` side of things as well (which would likely get ignored if/when things udpate, but hey it's better than nothing).

Reply:
I don't mind taking some more time to try and see how feasible it would be to do this refactor, either before or after this PR is merged. I will definitely add a comment to the original header pointing to the GPU code.